components/AuthProvider.tsx

```diff
@@ -0,0 +1,75 @@
+import { UserDoc } from 'models/users'
+import { onAuthStateChanged, User } from 'firebase/auth'
+import { doc, onSnapshot } from 'firebase/firestore'
+import React, {
+  createContext,
+  ReactNode,
+  useContext,
+  useEffect,
+  useState,
+} from 'react'
+import { auth, userRef } from 'utils/firebase'
+
+const authContext = createContext<{
+  user: User | null
+  userLoading: boolean
+  userData: UserDoc | null
+  userDataLoading: boolean
+}>({
+  user: null,
+  userLoading: true,
+  userData: null,
+  userDataLoading: true,
+})
+
+export const useUser = () => {
+  return useContext(authContext)
+}
+
+//There is a distinction between user and userData
+//User is the data used by Firebase to distinguish different user accounts
+//userData is the data we use to handle list ownership
+
+const AuthProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
+  const [user, setUser] = useState<User | null>(null)
+  const [userLoading, setLoading] = useState(true)
+  const [userData, setUserData] = useState<UserDoc | null>(null)
+  const [userDataLoading, setUserDataLoading] = useState(true)
+
+  useEffect(() => {
+    setUserDataLoading(true)
+    if (user) {
+      const unsub = onSnapshot(doc(userRef, user.uid), (doc) => {
+        const data = doc.data() as UserDoc
+        setUserData(data)
+        setUserDataLoading(false)
+      })
+
+      return unsub
+    } else {
+      setUserData(null)
+      setUserDataLoading(false)
+    }
+  }, [user])
```

Comment:
Two comments on this effect:

1. React's hooks work off of referential equality, so if `user` changes from on `User` object to another, this effect might not run as expected because I don't think the reference is actually changing. We have a package installed that has a `useDeepCompareEffect` if we wanna use object dependencies (though I might try and refactor it out at some point b/c imo we generally shouldn't need to use objects as dependencies - there's usually a property on the object we can check instead). See [this article](https://www.benmvp.com/blog/object-array-dependencies-react-useEffect-hook/) if you wanna get way too into React like I am 🤓 
2. **Prob the more important comment**, is this effect even necessary? We're already subscribing to auth changes below so can't we just put all the logic of this effect in there? We'd prob be able avoid the `user` check and having two loading state variables also.

Reply:
1. I ended up replacing the `useEffect `with `useDeepCompareEffect` but I don't think it would make much of a difference the users data that is stored on Firebase would rarely if ever change (this is the data that FireBase uses for auth), also if it re-ran I don't think we would encounter any issues, but feel free to let me know if I'm pretty curious.

2. I think it's necessary because it listens to changes in the user's `UserDoc `and not auth state changes which is what the `useEffect `below it does.  I'd keep it separate in case it needs to be re-run after auth state changes so in case a user signs out/ signs in with a different account we want to either create a new snapshot listener/ not depending on what the value of user is.

Reply:
1. Hmm, fair point that it prob won't change often - my concern was more that it might not run when the user changes, though I have ran into issues where effects repeatedly run because the dependency reference keeps changing even if the value is the same.
2. I was thinking something like this:
```
useEffect(() => {
    setLoading(true)

    let userDocUnsub: Unsubscribe
    const authUserUnsub = onAuthStateChanged(auth, (user) => {
      userDocUnsub = onSnapshot(doc(userRef, user?.uid), (doc) => {
        const data = doc.data() as UserDoc
        setUserData(data)
      })

      setUser(user)
      setLoading(false)
    })

    return () => {
      userDocUnsub()
      authUserUnsub()
    }
  }, [])
```
We're still listening to the user doc changes, just in the same effect as the auth user changes.